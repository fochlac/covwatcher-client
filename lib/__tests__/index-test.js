import { existsSync, readFile } from 'fs'

import FileHound from 'filehound'
import fetch from 'node-fetch'
import path from 'path'
import { upload } from '..'

jest.mock('filehound', () => {
	function FileHound() {
		return this
	}

	FileHound.create = () => {
		return FileHound
	}
	FileHound.paths = jest.fn(() => {
		return FileHound
	})
	FileHound.discard = () => {
		return FileHound
	}
	FileHound.match = () => {
		return FileHound
	}
	FileHound.find = jest
		.fn()
		.mockReturnValueOnce(Promise.resolve([]))
		.mockReturnValueOnce(Promise.resolve(['testpath']))

	return FileHound
})

jest.mock('fs', () => {
	return {
		existsSync: jest.fn(() => true),
		readFile: jest.fn((_a, _b, cb) => cb(null, 'testreport')),
	}
})

jest.mock('path', () => {
	return {
		join: jest.fn(() => '/test/path/for/this/test/'),
	}
})

jest.mock('node-fetch', () => jest.fn(() => Promise.resolve()))

jest.spyOn(process, 'exit').mockImplementation(val => {
	if (val !== 0) throw new Error('exit')
})

const params = [
	'',
	'',
	'-b',
	'-t',
	'branch',
	'-n',
	'name',
	'-y',
	'users',
	'-p',
	'project',
	'-r',
	'repo',
	'-s',
	'http://test.server.com:12345/',
	'-d',
	'/test/directory/clover.xml',
	'-l',
	'http://test.server.com:12345/lcovReport/',
	'-a',
]

describe('upload', () => {
	beforeEach(() => {
		jest.spyOn(global.console, 'log').mockImplementation(() => undefined)
	})
	afterEach(() => {
		global.console.log.mockRestore()
		fetch.mockClear()
		process.exit.mockClear()
		path.join.mockClear()
		existsSync.mockClear()
		readFile.mockClear()
		FileHound.find.mockClear()
		FileHound.paths.mockClear()
	})

	it('should print help and quit on -h', () => {
		process.argv = ['', '', '-h', '-b']
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing target', () => {
		process.argv = params.slice(0, 3)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if invalid target', () => {
		process.argv = params.slice(0, 4).concat(['invalidValue'])
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing name', () => {
		process.argv = params.slice(0, 5)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing type', () => {
		process.argv = params.slice(0, 7)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if invalid type', () => {
		process.argv = params.slice(0, 8).concat(['invalid'])
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing project', () => {
		process.argv = params.slice(0, 9)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing repo', () => {
		process.argv = params.slice(0, 11)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing repo', () => {
		process.argv = params.slice(0, 11)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should print error and quit if missing server', () => {
		process.argv = params.slice(0, 13)
		expect(() => upload()).toThrowError()
		expect(console.log).toMatchSnapshot()
	})

	it('should find test report and upload it to the server', async () => {
		process.argv = params.slice(0, 15)

		await upload()

		expect(FileHound.find).toBeCalledTimes(2)
		expect(FileHound.paths.mock.calls[1]).toMatchSnapshot()
		expect(existsSync).toBeCalledTimes(0)
		expect(readFile).toBeCalledTimes(1)

		expect(fetch.mock.calls).toMatchSnapshot()
	})

	it('should use passed in report and upload it to the server', async () => {
		process.argv = params.slice(0, 17)

		await upload()

		expect(FileHound.find).toBeCalledTimes(0)
		expect(existsSync).toBeCalledTimes(1)
		expect(readFile).toBeCalledTimes(1)

		expect(console.log).toMatchSnapshot()
		expect(fetch.mock.calls).toMatchSnapshot()
	})

	it('should also pass lcov prop to fetch', async () => {
		process.argv = params.slice(0, 19)

		await upload()

		expect(FileHound.find).toBeCalledTimes(0)
		expect(existsSync).toBeCalledTimes(1)
		expect(readFile).toBeCalledTimes(1)

		expect(console.log).toMatchSnapshot()
		expect(fetch.mock.calls).toMatchSnapshot()
	})

	it('should also pass task = true prop to fetch', async () => {
		process.argv = params.slice(0, 20)

		await upload()

		expect(FileHound.find).toBeCalledTimes(0)
		expect(existsSync).toBeCalledTimes(1)
		expect(readFile).toBeCalledTimes(1)

		expect(console.log).toMatchSnapshot()
		expect(fetch.mock.calls).toMatchSnapshot()
	})

	it('should go into catch on error', async () => {
		process.argv = params.slice(0, 17)
		readFile.mockImplementation((a, b, cb) => cb(true, null))
		let error
		try {
			await upload()
		} catch (err) {
			error = err
		}
		expect(error).toBeTruthy()
		expect(FileHound.find).toBeCalledTimes(0)
		expect(existsSync).toBeCalledTimes(1)
		expect(readFile).toBeCalledTimes(1)

		expect(console.log).toMatchSnapshot()
		expect(fetch).toBeCalledTimes(0)
	})
})
