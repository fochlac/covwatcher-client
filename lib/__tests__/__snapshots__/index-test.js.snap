// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload should find test report and upload it to the server 1`] = `
Array [
  "/test/path/for/this/test/",
]
`;

exports[`upload should find test report and upload it to the server 2`] = `
Array [
  Array [
    "http://test.server.com:12345//api/branch",
    Object {
      "name": "name",
      "report": "testreport",
      "repository": Object {
        "project": "project",
        "repo": "repo",
        "type": "users",
      },
    },
  ],
]
`;

exports[`upload should go into catch on error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error uploading coverage report:
",
      true,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if invalid target 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --target is neither 'branch' nor 'pullrequest'.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if invalid type 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --type is required. Please define the type of the repository. Can be either 'users' or 'project'.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing name 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --name is required. Please set the branchname / the id of the pullrequest.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing project 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --project is required. Please define the name of the project or the user the repository belongs to.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing repo 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --repo is required. Please define the name of the repository.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing repo 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --repo is required. Please define the name of the repository.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing server 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --server is required. Please define the url of the server.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing target 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --target is neither 'branch' nor 'pullrequest'.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print error and quit if missing type 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Error in command line arguments: --type is required. Please define the type of the repository. Can be either 'users' or 'project'.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should print help and quit on -h 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "	-n	--name		[name]			name of the branch or if this is a pullrequest its id
",
      "	-t	--target	[target]		branchtype, can be \\"branch\\" or \\"pullrequest\\"
",
      "	-y	--type		[type]			repository type, can be \\"users\\" or \\"project\\"
",
      "	-p	--project	[projectname]	name of the project or user the repository is part of
",
      "	-r	--repo		[repo]			name of the repository
",
      "	-s	--server	[server]		full url to the server running covwatcher. defaults to \\"localhost:51337\\"
",
      "	-d	--directory	[directory]		full path to the coverage report
",
      "	-b	--bail						repository type, can be \\"users\\" or \\"project\\"
",
      "	-h	--help						repository type, can be \\"users\\" or \\"project\\"
",
    ],
    Array [
      "Error in command line arguments: --target is neither 'branch' nor 'pullrequest'.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should use passed in report and upload it to the server 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Successfully uploaded coverage report.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`upload should use passed in report and upload it to the server 2`] = `
Array [
  Array [
    "http://test.server.com:12345//api/branch",
    Object {
      "name": "name",
      "report": "testreport",
      "repository": Object {
        "project": "project",
        "repo": "repo",
        "type": "users",
      },
    },
  ],
]
`;
